src/pathfinder/water_regions.cpp

```diff
@@ -0,0 +1,377 @@
+	{}
+
+	OrthogonalTileIterator begin() const { return this->tile_area.begin(); }
+	OrthogonalTileIterator end() const { return this->tile_area.end(); }
+
+	bool IsInitialized() const { return this->initialized; }
+
+	void Invalidate() { this->initialized = false; }
+
+	/**
+	 * Returns a set of bits indicating whether an edge tile on a particular side is traversable or not. These
+	 * values can be used to determine whether a ship can enter/leave the region through a particular edge tile.
+	 * @see GetLocalIndex() for a description of the coordinate system used.
+	 * @param side Which side of the region we want to know the edge traversability of.
+	 * @returns A value holding the edge traversability bits.
+	 */
+	TWaterRegionTraversabilityBits GetEdgeTraversabilityBits(DiagDirection side) const { return edge_traversability_bits[side]; }
+
+	/**
+	 * @returns The amount of individual water patches present within the water region. A value of
+	 * 0 means there is no water present in the water region at all.
+	 */
+	int NumberOfPatches() const { return this->number_of_patches; }
+
+	/**
+	 * @returns Whether the water region contains aqueducts that cross the region boundaries.
+	 */
+	bool HasCrossRegionAqueducts() const { return this->has_cross_region_aqueducts; }
+
+	/**
+	 * Returns the patch label that was assigned to the tile.
+	 * @param tile The tile of which we want to retrieve the label.
+	 * @returns The label assigned to the tile.
+	 */
+	TWaterRegionPatchLabel GetLabel(TileIndex tile) const
+	{
+		assert(this->tile_area.Contains(tile));
+		return this->tile_patch_labels[GetLocalIndex(tile)];
+	}
+
+	/**
+	 * Performs the connected component labeling and other data gathering.
+	 * @see WaterRegion
+	 */
+	void ForceUpdate()
+	{
+		this->has_cross_region_aqueducts = false;
+
+		this->tile_patch_labels.fill(INVALID_WATER_REGION_PATCH);
+
+		for (const TileIndex tile : this->tile_area) {
+			if (IsAqueductTile(tile)) {
+				const TileIndex other_aqueduct_end = GetOtherBridgeEnd(tile);
+				if (!tile_area.Contains(other_aqueduct_end)) this->has_cross_region_aqueducts = true;
+			}
+		}
+
+		TWaterRegionPatchLabel current_label = 1;
+		TWaterRegionPatchLabel highest_assigned_label = 0;
+
+		/* Perform connected component labeling. This uses a flooding algorithm that expands until no
+		 * additional tiles can be added. Only tiles inside the water region are considered. */
+		for (const TileIndex start_tile : tile_area) {
+			static std::vector<TileIndex> tiles_to_check;
+			tiles_to_check.clear();
+			tiles_to_check.push_back(start_tile);
+
+			bool increase_label = false;
+			while (!tiles_to_check.empty()) {
+				const TileIndex tile = tiles_to_check.back();
+				tiles_to_check.pop_back();
+
+				const TrackdirBits valid_dirs = TrackBitsToTrackdirBits(GetWaterTracks(tile));
+				if (!valid_dirs) continue;
```

Comment:
```suggestion
				if (valid_dirs == TRACKDIR_BIT_NONE) continue;
```
The `!` is generally only allowed for booleans.